encoders/firebase-decoders-json/src/json/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContext.java

```diff
@@ -145,6 +148,83 @@
     return null;
   }
 
+  private <T> boolean isPreDefinedObject(TypeToken.ClassToken<T> classToken) {
+    Class<T> clazz = classToken.getRawType();
+    return List.class.isAssignableFrom(clazz)
+        || Map.class.isAssignableFrom(clazz)
+        || Set.class.isAssignableFrom(clazz);
+  }
+
+  private <T> T decodePreDefinedObject(TypeToken.ClassToken<T> classToken) throws IOException {
+    Class<T> clazz = classToken.getRawType();
+    if (Map.class.isAssignableFrom(clazz)) {
+      @SuppressWarnings("unchecked")
+      T t = (T) decodeMap((TypeToken.ClassToken<Map>) classToken);
+      return t;
+    } else if (List.class.isAssignableFrom(clazz)) {
+      // TODO: support List
+    } else if (Set.class.isAssignableFrom(clazz)) {
+      // TODO: support List
+    }
+    return null;
+  }
+
+  private <K, V, T extends Map<K, V>> T decodeMap(TypeToken.ClassToken<T> classToken)
+      throws IOException {
+    TypeToken<K> keyTypeToken = classToken.getTypeArguments().at(0);
+    TypeToken<V> valTypeToken = classToken.getTypeArguments().at(1);
+    if (!isSingleValue(keyTypeToken))
+      throw new IllegalArgumentException(keyTypeToken + " cannot be used as Map key.");
+
+    T map = null;
+    try {
+      map = classToken.getRawType().getDeclaredConstructor().newInstance();
+    } catch (Exception e) {
+      throw new IllegalArgumentException(
+          classToken.getRawType()
+              + " cannot be initialized.\n"
+              + "Do not pass abstract class and an interface.\n"
+              + e);
+    }
```

Comment:
Are there better ways to deal with Exceptions?
or avoid using reflect?

Reply:
We don't want to use reflection, I would say let's start with supporting `HashMap` and `LinkedHashMap` and add support for any others later if we have to.

In general users should be encouraged to have fields of interface type `Map` not concrete implementations, so I think supporting `Map`, `HashMap` and `LinkedHashMap` should be good enough.